### PR TITLE
Assign error codes to all of our messages

### LINKED ISSUE #2065: Add codes to MSBuild warnings and errors
Follow-up to this: https://github.com/dotnet/sdk/pull/2064#discussion_r175557190

Most MSBuild warnings and errors issued by this SDK do not have error codes. We should add codes to all of them.

### PR DESCRIPTION
**Pull request by nguerrera**
Fix #2065 

This adds "NETSDK1234" error codes to each of our error/warning/log messages and sends them to msbuild as such.

We need to add a new message with a code in order for project system to light up an info bar. This puts the infrastructure in place for that.

Individual commits may be easier to review.

The most interesting commits are:
* 9596e8d - Shows the pattern for adding a new error code to resx. From there we can call Log.LogXxx as before.
* d3ffdf8 - Implements the plumbing of the codes from resx to msbuild.

### CODE REVIEW

**Review by livarcocc — APPROVED** [review]

File: src/Tasks/Common/Resources/Strings.resx [hunk 1]
```diff
@@ -118,202 +118,267 @@
     <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
   </resheader>
   <data name="AtLeastOneTargetFrameworkMustBeSpecified" xml:space="preserve">
-    <value>At least one possible target framework must be specified.</value>
+    <value>NETSDK1001: At least one possible target framework must be specified.</value>
```
Comment (dsplaisted) [inline src/Tasks/Common/Resources/Strings.resx]: Let's add tests that validate that we don't duplicate error codes over time.

To ensure that this still holds even for deleted messages, we discussed having a list of deleted message codes, and ensuring that there are no duplicates between the current and deleted message codes, and that the current and deleted message codes are contiguous.

File: src/Tasks/Common/Logger.cs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    ///
+    /// Extracts error codes from the message prefix, and enforces that all of
+    /// our messages have a NETSDK code.
+    ///
+    /// Example:
+    ///   C#
+    ///     Log.LogError(Strings.SomethingIsWrong);
+    ///
+    ///   Strings.resx:
+    ///     | SomethingIsWrong | NETSDK1234: Something is wrong. |
```
Comment (dsplaisted) [inline src/Tasks/Common/Logger.cs]: The pipes didn't make it clear to me that this wasn't just one long string.  Perhaps something like this would be clearer:

```
  Strings.resx:
    Resource Name: SomethingIsWrong
    Resource Value: NETSDK1234: Something is wrong.
```

File: src/Tasks/Microsoft.NET.Build.Tasks/LockFileCache.cs [hunk 3]
```diff
@@ -107,28 +107,26 @@ private LockFile LoadLockFile(string path)
...
-                        break;
+                    throw new BuildErrorException(
+                        string.Format(Strings.ErrorReadingAssetsFile, message.Message));
                 }
+
+                _log.Log(
+                    new Message(
+                        level: message.Level == LogLevel.Warning ? MessageLevel.Warning : MessageLevel.NormalImportance,
+                        code: message.Code == NuGetLogCode.Undefined ? default : message.Code.ToString(),
+                        text: message.Message));
```
Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/LockFileCache.cs]: Plumb the file through here from the NuGet message.

File: src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenALogger.cs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+            logger.LogError("NETSDK0000: _");
+            logger.HasLoggedErrors.Should().BeTrue();
+        }
+
+        [Fact]
+        public void ItEnforcesErrorCodesInDebug()
+        {
+            var logger = new TestLogger();
+
+            logger.LogError("No error code");
```
Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenALogger.cs]: Won't this fail in Debug?  I think it needs to be moved into the `#else` clause, or possibly delete it, bring `logWithoutErrorCode` out of the `#if` clause, and invoke it in the `#else` clause.

**Review by dsplaisted — APPROVED** [review]
Looks good, please add tests to ensure we don't duplicate message codes as discussed.

File: src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenALogger.cs (hunk 4 above)
Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenALogger.cs]: Oops.

**Comment by nguerrera** [comment]
> please add tests to ensure we don't duplicate message codes as discussed.

@dsplaisted Done in b038a28

The other commits since review are fixups to be rebased away that 1) fix bugs found by the test (yay!) 2) adress other minor feedback.

**Review by dsplaisted — APPROVED** [review]

File: src/Tasks/Common/Resources/Strings.resx (hunk 1 above)
Comment (nguerrera) [reply src/Tasks/Common/Resources/Strings.resx]: Done.
